### ISSUE TITLE
feature request: Siamese Neural Network

### ISSUE THREAD

**Issue by MechaCritter** [MechaCritter/Python-Visual-Similarity#27]
**Is your feature request related to a problem? Please describe.**
I would like to implement the class **SiameseNeuralNetwork**. It is currently under `pyvisim/neural_networks`. If you are new to the concept of Siamese Neural Net, you can check out [this link here](https://dvl.in.tum.de/slides/adl4cv-ss20/2.Siamese.pdf).

If you're new to the codebase, feel free to read the [technical documentation here](https://github.com/MechaCritter/Python-Visual-Similarity/blob/main/docs/overview.md).

**PLEASE**: implementing this class requires understanding of deep learning and computer vision. Please do not commit random vibe-coded stuff without checking, as I will then reject those PRs. Please save me and save yourself time :)

**Describe the solution you'd like**
I defined a sketch for this class [here](https://github.com/MechaCritter/Python-Visual-Similarity/blob/main/pyvisim/neural_networks/siamese_neural_network.py). This is, however, only a sketch. If you have a better idea on how to implement this class, feel free to show your idea. But make sure it's a trainable `torch.nn.Module` network so that it can be trained properly within the PyTorch framework.

However, methods like `encode` and `similarity_score` remains, as these methods provide this class a similar interface with the rest of the codebase.

If you have time, you can research and come up with a way to train this neural net, and then provide the training guides in the docs [here](https://github.com/MechaCritter/Python-Visual-Similarity/blob/main/docs/neural_networks/README.md). Also, under `tests/`, you can write a small test suite where you train the neural net on a subsample of images from the **OxfordFlowerDataset** (under `pyvisim/datasets`, see the [docs here](https://github.com/MechaCritter/Python-Visual-Similarity/blob/main/docs/dataset/oxford_flower_dataset.md)) to see how the network performs when distinguishing between flowers of different labels.

**Describe alternatives you've considered**
If you have a better idea, feel free to come up with a completely different neural network. If you implement a network from a paper or from another GitHub repository, please cite the source.

**Additional context**
- Siamese Neural Net literature: 
+ https://dvl.in.tum.de/slides/adl4cv-ss20/2.Siamese.pdf
+ https://medium.com/@rinkinag24/a-comprehensive-guide-to-siamese-neural-networks-3358658c0513
